README.md
# extended_theme

Easy management of themes for Flutter apps. Create theme sets, extend standard themes, update themes in runtime.
Define your theme classes or use Material and Cupertino themes out the box! 

![](.assets/screencast_small.gif)

## Usage

### Using predefined themes (skins)
1. Define your theme class that extends `ExtendedTheme` (Optional, bacause you may to use class `ExtendedTheme` out the box) Do not forget call super constructor and pass them `material` or `cupertino` theme data.
```dart
class AppTheme extends ExtendedTheme {
  final Color? shadowColor;
  final Color? buttonPauseColor;
  final Color? subtitleColor;
  final double? centerFontSize;

  AppTheme(ThemeData material,
      {this.shadowColor,
      this.buttonPauseColor,
      this.subtitleColor,
      this.centerFontSize})
      : super(material: material);
}

```
2. Define themes configuration (skins)
```dart
const MagentaLight = 'MagentaLight';
const MagentaDark = 'MagentaDark';
const GreenLight = 'GreenLight';
const GreenDark = 'GreenDark';
const BlueLight = 'BlueLight';
const BlueDark = 'BlueDark';

final appThemes = {
  MagentaLight: AppTheme(
      ThemeData.light().copyWith(primaryColor: const Color(0xffcc0066)),
      subtitleColor: const Color(0xff8e8e8e),
      centerFontSize: 28),

  MagentaDark: AppTheme(
      ThemeData.dark().copyWith(primaryColor: const Color(0xffcc0066)),
      subtitleColor: const Color(0xffA5A5A5),
      centerFontSize: 14),

  GreenLight: AppTheme(ThemeData.light().copyWith(primaryColor: Colors.green),
      subtitleColor: const Color(0xff8e8e8e), centerFontSize: 22),

  GreenDark: AppTheme(ThemeData.dark().copyWith(primaryColor: Colors.green[700]),
      subtitleColor: const Color(0xffA5A5A5), centerFontSize: 12),

  BlueLight: AppTheme(ThemeData.light().copyWith(primaryColor: Colors.blue),
      subtitleColor: const Color(0xff8e8e8e), centerFontSize: 14),

  BlueDark: AppTheme(ThemeData.dark().copyWith(primaryColor: Colors.blue[700]),
      subtitleColor: const Color(0xffA5A5A5), centerFontSize: 10) //
};

```

3. Wrap your root widget in `ThemeScope` widget and link material theme with needed property of material app. Do not forget to use `themeBuilder`
```dart
class MyApp extends StatelessWidget {
  // Read this value from SharedPrefs for example
  final String initialTheme;

  const MyApp({Key key, required this.initialTheme}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ThemeScope<AppTheme>(
      themeId: initialTheme,
      // Skins that your defined on previous step
      availableThemes: appThemes,
      themeBuilder: (context, appTheme) {
        return MaterialApp(
          title: 'Flutter Demo',
          // Set material Theme Data
          theme: appTheme.material,
          home: MyHomePage(title: 'Flutter Extended Theme'),
        );
      },
    );
  }
}

```
4. Getting theme properties in your widget tree. By `ThemeHolder`:
```dart
ThemeHolder.themeOf<AppTheme>(context).centerFontSize

```
Or by `Theme` (if you use material app and no need access to additional properties):
```dart
Theme.of(context).textTheme.headline4

```

5. Update theme (skin):
```dart
ThemeHolder.of<AppTheme>(context)
              .updateThemeById('MagentaDark');

```


### Build themes in runtime
Sometimes the use of predefined themes imposes a limitation. For example, if you want to keep the users selected font separately. This is done by building the themes in the runtime.

1. Define your class extends `ExtendedTheme` (Optional, performed in the same way as in the previous section) 
2. Build your initial theme and pass it to `ThemeScope` constructor:
```dart
class MyApp extends StatelessWidget {
  final AppTheme initialTheme = AppTheme(ThemeData.light(), centerFontSize: 22, 
    subtitleColor: const Color(0xff8e8e8e));

  MyApp({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ThemeScope<AppTheme>(
      theme: initialTheme,
      themeBuilder: (context, appTheme) {
        return MaterialApp(
          title: 'Flutter Demo',
          // Set material Theme Data
          theme: appTheme.material,
          home: MyHomePage(title: 'Flutter Extended Theme'),
        );
      },
    );
  }
}

```

3. Update theme in runtime
```dart
// Build new theme
final newTheme = AppTheme(ThemeData.dark(), centerFontSize: 12, 
    subtitleColor: const Color(0xff8e8e8e));

// Update theme 
ThemeHolder.of<AppTheme>(context).updateTheme(newTheme);

```

4. Getting theme properties in your widget tree. Performed in the same way as in the previous section.

### Work with brightness

You can control light and dark themes, but you need to define the appropriate property in your theme class. For example:

```dart
class AppTheme extends ExtendedTheme {
  final ThemeData materialDark;

  AppTheme(ThemeData material, this.materialDark)
      : super(material: material);
}

```

Then pass both themes to MaterialApp constructor:
```dart
class MyApp extends StatelessWidget {
  final String initialTheme;

  const MyApp({Key key, required this.initialTheme}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ThemeScope<AppTheme>(
      themeId: initialTheme,
      availableThemes: appThemes,
      themeBuilder: (context, appTheme) {
        return MaterialApp(
          title: 'Flutter Demo',
          // Set default (light) theme data
          theme: appTheme.material,
          // Set dark theme data
          darkTheme: appTheme.materialDark,
          home: MyHomePage(title: 'Flutter Extended Theme'),
        );
      },
    );
  }
}

```

CHANGELOG.md
## [0.2.0]
### Update to null safety

## [0.1.1]

### Enhances
- Update description
## [0.1.0] 

### Initial release

test/test_themes.dart
import 'dart:ui';

import 'package:extended_theme/extended_theme.dart';
import 'package:flutter/material.dart';

class TestTheme extends ExtendedTheme {
  final Color? shadowColor;
  final Color? buttonPauseColor;
  final Color? subtitleColor;

  TestTheme(ThemeData material,
      {this.shadowColor, this.buttonPauseColor, this.subtitleColor})
      : super(material: material);
}

ThemeData baseTheme = ThemeData(
  brightness: Brightness.light,
);

ThemeData baseThemeDark = ThemeData(
  brightness: Brightness.dark,
);

const MagentaLight = 'MagentaLight';
const MagentaDark = 'MagentaDark';
const GreenLight = 'GreenLight';
const GreenDark = 'GreenDark';
const BlueLight = 'BlueLight';
const BlueDark = 'BlueDark';

final appThemes = {
  MagentaLight: TestTheme(
      baseTheme.copyWith(primaryColor: const Color(0xffcc0066)),
      subtitleColor: const Color(0xff8e8e8e)),

  MagentaDark: TestTheme(
      baseThemeDark.copyWith(primaryColor: const Color(0xffcc0066)),
      subtitleColor: const Color(0xffA5A5A5)),

  GreenLight: TestTheme(baseTheme.copyWith(primaryColor: Colors.green),
      subtitleColor: const Color(0xff8e8e8e)),

  GreenDark: TestTheme(baseThemeDark.copyWith(primaryColor: Colors.green[700]),
      subtitleColor: const Color(0xffA5A5A5)),

  BlueLight: TestTheme(
    baseTheme.copyWith(primaryColor: Colors.blue),
    subtitleColor: const Color(0xff8e8e8e),
  ),

  BlueDark: TestTheme(baseThemeDark.copyWith(primaryColor: Colors.blue[700]),
      subtitleColor: const Color(0xffA5A5A5)) //
};


test/test_app_runtime.dart
import 'package:extended_theme/extended_theme.dart';
import 'package:flutter/material.dart';

import 'test_themes.dart';

class TestAppRuntime extends StatelessWidget {
  final TestTheme? initialTheme;

  const TestAppRuntime({Key? key, this.initialTheme}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ThemeScope<TestTheme>(
      theme: initialTheme,
      themeBuilder: (context, appTheme) {
        return MaterialApp(
          title: 'Flutter Extended Theme Test (Runtime)',
          theme: appTheme.material,
          home: Container(
            color:
                ThemeHolder.themeOf<TestTheme>(context).material!.accentColor,
            child: Column(
              children: [
                Center(
                    child: TextButton(
                  onPressed: () {
                    final newTheme = TestTheme(
                        ThemeData(accentColor: Colors.red),
                        subtitleColor: Colors.yellow);

                    ThemeHolder.of<TestTheme>(context).updateTheme(newTheme);
                  },
                  child: Text('Change theme'),
                )),
                Text(
                  ThemeHolder.of<TestTheme>(context)
                      .theme
                      .subtitleColor!
                      .value
                      .toString(),
                  key: ValueKey('themeIdText'),
                ),
              ],
            ),
          ),
        );
      },
    );
  }
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp(initialTheme: GreenLight));

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/lib/theme.dart
import 'package:extended_theme/extended_theme.dart';
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';

class AppTheme extends ExtendedTheme {
  final Color shadowColor;
  final Color buttonPauseColor;
  final Color subtitleColor;
  final double centerFontSize;

  AppTheme(ThemeData material, this.centerFontSize,
      {this.shadowColor, this.buttonPauseColor, this.subtitleColor})
      : super(material: material);
}

ThemeData baseTheme = ThemeData(
  brightness: Brightness.light,
  textTheme: GoogleFonts.openSansTextTheme(ThemeData.light().textTheme),
);

ThemeData baseThemeDark = ThemeData(
  brightness: Brightness.dark,
  textTheme: GoogleFonts.openSansTextTheme(ThemeData.dark().textTheme),
);

const MagentaLight = 'MagentaLight';
const MagentaDark = 'MagentaDark';
const GreenLight = 'GreenLight';
const GreenDark = 'GreenDark';
const BlueLight = 'BlueLight';
const BlueDark = 'BlueDark';

final appThemes = {
  MagentaLight: AppTheme(
      baseTheme.copyWith(primaryColor: const Color(0xffcc0066)), 28,
      subtitleColor: const Color(0xff8e8e8e)),

  MagentaDark: AppTheme(
      baseThemeDark.copyWith(primaryColor: const Color(0xffcc0066)), 14,
      subtitleColor: const Color(0xffA5A5A5)),

  GreenLight: AppTheme(baseTheme.copyWith(primaryColor: Colors.green), 22,
      subtitleColor: const Color(0xff8e8e8e)),

  GreenDark: AppTheme(
      baseThemeDark.copyWith(primaryColor: Colors.green[700]), 12,
      subtitleColor: const Color(0xffA5A5A5)),

  BlueLight: AppTheme(
    baseTheme.copyWith(primaryColor: Colors.blue),
    14,
    subtitleColor: const Color(0xff8e8e8e),
  ),

  BlueDark: AppTheme(baseThemeDark.copyWith(primaryColor: Colors.blue[700]), 10,
      subtitleColor: const Color(0xffA5A5A5)) //
};


lib/extended_theme.dart
library extended_theme;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

typedef ThemedWidgetBuilder<TTheme extends ExtendedTheme> = Widget Function(
    BuildContext context, TTheme theme);

/// Base class for themes.
/// You can use this class directly without creating any descendants
class ExtendedTheme {
  final ThemeData? material;
  final CupertinoThemeData? cupertino;

  @mustCallSuper
  ExtendedTheme({this.material, this.cupertino})
      : assert(material != null || cupertino != null);

  ExtendedTheme copyWith({
    ThemeData? material,
    CupertinoThemeData? cupertino,
  }) {
    return ExtendedTheme(
      material: material ?? this.material,
      cupertino: cupertino ?? this.cupertino,
    );
  }

  @override
  bool operator ==(Object o) {
    if (identical(this, o)) return true;

    return o is ExtendedTheme &&
        o.material == material &&
        o.cupertino == cupertino;
  }

  @override
  int get hashCode => material.hashCode ^ cupertino.hashCode;
}

/// Widget for managing the application themes
/// Wrap in it your root widget to manage the application theme
class ThemeScope<TTheme extends ExtendedTheme> extends StatefulWidget {
  /// Defines the original theme from which your application will be started.
  /// If you use this field, you should also define the theme map - [availableThemes]
  final String? themeId;

  /// Predefined themes (skins) for you app
  final Map<String, TTheme>? availableThemes;

  /// Initial theme if you do not want to use predefined themes
  final TTheme? theme;

  /// Root widget builder
  final ThemedWidgetBuilder<TTheme> themeBuilder;

  const ThemeScope(
      {Key? key,
      this.themeId,
      this.availableThemes,
      this.theme,
      required this.themeBuilder})
      : super(key: key);

  @override
  _ThemeScopeState<TTheme> createState() => _ThemeScopeState<TTheme>();
}

/// A controller that stores a link to [theme] and allows you to update it
class ThemeHolder<TTheme extends ExtendedTheme> {
  final _ThemeScopeState<TTheme> _facilityState;

  ThemeHolder(this._facilityState);

  /// Current theme identifier if you use predefined map of themes
  /// Equals null if you update theme in runtime by method [updateTheme]
  String? get themeId => _facilityState.themeId;

  /// Current theme. Use this in the Widget tree for getting of theme properties
  TTheme get theme => _facilityState.theme;

  /// Update the theme by the identifier that is
  /// contained in the map of themes you have defined
  void updateThemeById(String themeId) {
    _facilityState.updateThemeById(themeId);
  }

  /// Update the theme by theme object
  void updateTheme(TTheme theme) {
    _facilityState.updateTheme(theme);
  }

  /// Link to theme controller object, use this for managing themes
  static ThemeHolder<TTheme> of<TTheme extends ExtendedTheme>(
      BuildContext context) {
    return context
        .dependOnInheritedWidgetOfExactType<_InheritedTheme<TTheme>>()!
        .stateTheme!
        .themeFacade;
  }

  /// Current theme, use this in the Widget tree for getting theme properties
  static TTheme themeOf<TTheme extends ExtendedTheme>(BuildContext context) {
    return ThemeHolder.of<TTheme>(context).theme;
  }
}

class _ThemeScopeState<TTheme extends ExtendedTheme>
    extends State<ThemeScope<TTheme>> {
  String? _themeId;
  TTheme? _theme;

  late ThemeHolder<TTheme> _facade;

  String? get themeId => _themeId;
  TTheme get theme {
    if (_themeId != null) {
      return widget.availableThemes![_themeId!]!;
    }

    return _theme!;
  }

  ThemeHolder<TTheme> get themeFacade => _facade;

  void _initData() {
    if (widget.themeId != null && widget.theme != null) {
      throw Exception(
          'It is not allowed to specify both the identifier and an instance of the theme!');
    }

    if (widget.themeId != null) {
      if (widget.availableThemes == null) {
        throw Exception(
            'When passing an identifier, you need to specify the map of themes!');
      } else {
        _checkThemeId(widget.themeId);
      }
    }

    _themeId = widget.themeId;
    _theme = widget.theme;
  }

  void _checkThemeId(String? newThemeId) {
    if (!widget.availableThemes!.containsKey(newThemeId)) {
      throw Exception('Themes does not contains this key!');
    }
  }

  void updateThemeById(String newThemeId) {
    if (_themeId != newThemeId) {
      _checkThemeId(newThemeId);

      setState(() {
        _themeId = newThemeId;
        //debugPrint('Updated theme: ' + newThemeId.toString());
      });
    }
  }

  void updateTheme(TTheme theme) {
    setState(() {
      _themeId = null;
      _theme = theme;
    });
  }

  @override
  void initState() {
    super.initState();

    _facade = ThemeHolder(this);
    _initData();
  }

  @override
  void didUpdateWidget(ThemeScope oldWidget) {
    super.didUpdateWidget(oldWidget as ThemeScope<TTheme>);
    _initData();
  }

  @override
  Widget build(BuildContext context) {
    return _InheritedTheme<TTheme>(
      stateTheme: this,
      child: Builder(builder: (themeContext) {
        return widget.themeBuilder(themeContext, theme);
      }),
    );
  }
}

class _InheritedTheme<TTheme extends ExtendedTheme> extends InheritedWidget {
  final _ThemeScopeState<TTheme>? stateTheme;

  _InheritedTheme({Key? key, this.stateTheme, required Widget child})
      : super(key: key, child: child);

  @override
  bool updateShouldNotify(covariant _InheritedTheme<TTheme> oldWidget) {
    return oldWidget.stateTheme!.themeId != stateTheme!.themeId ||
        oldWidget.stateTheme!.theme != stateTheme!.theme; //true;
  }
}


example/lib/main.dart
import 'package:extended_theme/extended_theme.dart';
import 'package:flutter/material.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'package:google_fonts/google_fonts.dart';

class AppTheme extends ExtendedTheme {
  final Color? shadowColor;
  final Color? buttonPauseColor;
  final Color? subtitleColor;
  final double? centerFontSize;

  AppTheme(ThemeData material,
      {this.shadowColor,
      this.buttonPauseColor,
      this.subtitleColor,
      this.centerFontSize})
      : super(material: material);
}

ThemeData baseTheme = ThemeData(
  brightness: Brightness.light,
  textTheme: GoogleFonts.openSansTextTheme(ThemeData.light().textTheme),
);

ThemeData baseThemeDark = ThemeData(
  brightness: Brightness.dark,
  textTheme: GoogleFonts.openSansTextTheme(ThemeData.dark().textTheme),
);

const MagentaLight = 'MagentaLight';
const MagentaDark = 'MagentaDark';
const GreenLight = 'GreenLight';
const GreenDark = 'GreenDark';
const BlueLight = 'BlueLight';
const BlueDark = 'BlueDark';

final appThemes = {
  MagentaLight: AppTheme(
      baseTheme.copyWith(primaryColor: const Color(0xffcc0066)),
      subtitleColor: const Color(0xff8e8e8e),
      centerFontSize: 28),

  MagentaDark: AppTheme(
      baseThemeDark.copyWith(primaryColor: const Color(0xffcc0066)),
      subtitleColor: const Color(0xffA5A5A5),
      centerFontSize: 14),

  GreenLight: AppTheme(baseTheme.copyWith(primaryColor: Colors.green),
      subtitleColor: const Color(0xff8e8e8e), centerFontSize: 22),

  GreenDark: AppTheme(baseThemeDark.copyWith(primaryColor: Colors.green[700]),
      subtitleColor: const Color(0xffA5A5A5), centerFontSize: 12),

  BlueLight: AppTheme(baseTheme.copyWith(primaryColor: Colors.blue),
      subtitleColor: const Color(0xff8e8e8e), centerFontSize: 14),

  BlueDark: AppTheme(baseThemeDark.copyWith(primaryColor: Colors.blue[700]),
      subtitleColor: const Color(0xffA5A5A5), centerFontSize: 10) //
};

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  /// Read last theme from [SharedPreferences] for simple example
  /// Avoid [async] in main method for production. Try to use FutureBuilder or other async widgets
  final initialTheme = await _readLastTheme();

  runApp(MyApp(
    initialTheme: initialTheme,
  ));
}

Future<String> _readLastTheme() async {
  var prefs = await SharedPreferences.getInstance();

  final themeStr = prefs.getString('PREF_THEME') ?? GreenLight;
  return themeStr;
}

class MyApp extends StatelessWidget {
  final String initialTheme;

  const MyApp({Key? key, required this.initialTheme}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ThemeScope<AppTheme>(
      themeId: initialTheme,
      availableThemes: appThemes,
      themeBuilder: (context, appTheme) {
        debugPrint('Builder build');

        return MaterialApp(
          title: 'Flutter Demo',
          theme: appTheme.material,
          home: MyHomePage(title: 'Flutter Extended Theme'),
        );
      },
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);
  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have changed the theme many times:',
              textAlign: TextAlign.center,
              style: TextStyle(
                  fontSize:
                      ThemeHolder.themeOf<AppTheme>(context).centerFontSize),
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        backgroundColor:
            ThemeHolder.themeOf<AppTheme>(context).buttonPauseColor,
        onPressed: () {
          final themeKeys = appThemes.keys.toList();
          ThemeHolder.of<AppTheme>(context)
              .updateThemeById(themeKeys[_counter % themeKeys.length]);

          _incrementCounter();
        },
        tooltip: 'Next theme',
        child: Icon(Icons.colorize),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
